packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -952,104 +993,106 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         }
 
         tryCancel();
-
+        
         if (!request) return;
 
-        // Reset loadedUrl so an old one is not carried over to retries.
-        request.loadedUrl = undefined;
-
-        const statisticsId = request.id || request.uniqueKey;
-        this.stats.startJob(statisticsId);
-
-        // Shared crawling context
-        // @ts-expect-error
-        // All missing properties properties (that extend CrawlingContext) are set dynamically,
-        // but TS does not know that, so otherwise it would throw when compiling.
-        const crawlingContext: Context = {
-            id: cryptoRandomObjectId(10),
-            crawler: this,
-            log: this.log,
-            request,
-            session,
-            enqueueLinks: async (options: SetRequired<EnqueueLinksOptions, 'urls'>) => {
-                return enqueueLinks({
-                    // specify the RQ first to allow overriding it
-                    requestQueue: await this.getRequestQueue(),
-                    ...options,
-                });
-            },
-            sendRequest: async (overrideOptions?: OptionsInit) => {
-                const cookieJar = session ? {
-                    getCookieString: async (url: string) => session!.getCookieString(url),
-                    setCookie: async (rawCookie: string, url: string) => session!.setCookie(rawCookie, url),
-                    ...overrideOptions?.cookieJar,
-                } : overrideOptions?.cookieJar;
-
-                return gotScraping({
-                    url: request!.url,
-                    method: request!.method as Method, // Narrow type to omit CONNECT
-                    body: request!.payload,
-                    headers: request!.headers,
-                    proxyUrl: crawlingContext.proxyInfo?.url,
-                    sessionToken: session,
-                    responseType: 'text',
-                    ...overrideOptions,
-                    retry: {
-                        limit: 0,
-                        ...overrideOptions?.retry,
-                    },
-                    cookieJar,
-                });
-            },
-        };
-
-        this.crawlingContexts.set(crawlingContext.id, crawlingContext);
-
-        try {
-            request.state = RequestState.REQUEST_HANDLER;
-            await addTimeoutToPromise(
-                () => this._runRequestHandler(crawlingContext),
-                this.requestHandlerTimeoutMillis,
-                `requestHandler timed out after ${this.requestHandlerTimeoutMillis / 1000} seconds (${request.id}).`,
-            );
-
-            await this._timeoutAndRetry(
-                () => source.markRequestHandled(request!),
-                this.internalTimeoutMillis,
-                `Marking request ${request.url} (${request.id}) as handled timed out after ${this.internalTimeoutMillis / 1e3} seconds.`,
-            );
+        if(this._handleRequestWithDelay(request, source)){
```

Comment:
same as above, this should check the opposite instead of nesting the large block of code more. its really hard to see what you changed because of this